### PR TITLE
Bug 2226 fast replay streaming txmeta

### PR DESCRIPTION
**Pull request by graydon**
# Description

First cut at #2226. This has 3 main pieces:

  - A new XDR type of high-resolution metadata called `LedgerCloseMeta` that captures a per-txn, per-ledger-entry change stream that's equivalent to / a superset of the history tables currently being written into the core database for horizon's benefit.
  - A new "application mode" field in ApplicationImpl that lets us switch-off subsystems and/or replace them with in-memory stubs when running in modes that don't use them.
  - A new catchup option `--replay-in-memory` that runs catchup in application mode `REPLAY_HISTORY_FOR_METADATA`, that has no overlay or database, and uses a "stub" `LedgerTxnRoot` that returns empty for any query, and a "never committed" pseudo-root `LedgerTxn` connected to it, that we use as the in-memory ledger.

The idea here is for horizon to write a config file that mentions a local stream and then runs stellar-core as a subprocess, in either bulk replay mode (`catchup --replay-in-memory`) during initial ingestion _or_ in `run` mode to track the ongoing changes in the network.

(cc @ire-and-curses and @bartekn)

### CODE REVIEW

File: src/ledger/LedgerManagerImpl.h [hunk 1]
```diff
@@ -67,11 +71,13 @@ class LedgerManagerImpl : public LedgerManager
                          CatchupConfiguration::Mode catchupMode);
 
     void processFeesSeqNums(std::vector<TransactionFramePtr>& txs,
-                            AbstractLedgerTxn& ltxOuter, int64_t baseFee);
+                            AbstractLedgerTxn& ltxOuter, int64_t baseFee,
+                            LedgerCloseMeta* ledgerCloseMeta);
```
Comment (jonjove) [inline src/ledger/LedgerManagerImpl.h]: Why take the raw-pointer as argument instead of `std::unique_ptr<LedgerCloseMeta> const&`? Passing the smart-pointer makes ownership questions (eg. do I have to delete this?) much clearer.

File: src/herder/TxSetFrame.cpp [hunk 2]
```diff
@@ -512,6 +513,8 @@ TxSetFrame::getTotalFees(LedgerHeader const& lh) const
 void
 TxSetFrame::toXDR(TransactionSet& txSet)
 {
+    // Should only call toXDR when in sorted-for-hash state.
+    releaseAssertOrThrow(mHashIsValid);
```
Comment (marta-lokhova) [inline src/herder/TxSetFrame.cpp]: I don't think we can guarantee this currently:  some places like `StateSnapshot::writeHistoryBlocks`, or `HerderImpl::persistSCPState` convert to XDR without necessarily calculating the hash (which sets mHashIsValid to true)

File: src/main/Application.h [hunk 3]
```diff
@@ -157,6 +181,33 @@ class Application
...
 
+    enum class AppMode
+    {
+        // Application will connect to networks (unless Cfg.RUN_STANDALONE),
+        // bind to a database and bucket list, and receive and apply
+        // consensus transactions. "Normal operation".
+        RUN_LIVE_NODE,
+
+        // Application will connect to networks but _not_ establish a database
+        // of any sort, nor participate in consensus, apply transactions or
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/main/Application.h]: I think consensus, or at least voting in it, is an optional aspect of this mode -- in the sense that we could also set the `NODE_IS_VALIDATOR` flag to false -- but I'll change the comment to reflect that.

File: src/main/Maintainer.cpp [hunk 4]
```diff
@@ -28,6 +28,10 @@ Maintainer::start()
 void
 Maintainer::scheduleMaintenance()
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/main/Maintainer.cpp]: Good catch. fixed.

File: src/ledger/InMemoryLedgerTxnRoot.cpp [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+InMemoryLedgerTxnRoot::getInflationWinners(size_t maxWinners,
+                                           int64_t minBalance)
+{
+    return std::vector<InflationWinner>();
+}
+
+std::shared_ptr<LedgerEntry const>
+InMemoryLedgerTxnRoot::getNewestVersion(LedgerKey const& key) const
+{
+    return nullptr;
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/ledger/InMemoryLedgerTxnRoot.cpp]: There's some mention in ApplicationImpl.h but I'll expand it a bit.

File: src/main/Application.cpp [hunk 6]
```diff
@@ -37,8 +37,51 @@ validateNetworkPassphrase(Application::pointer app)
...
-    return create<ApplicationImpl>(clock, cfg, newDB);
+    return create<ApplicationImpl>(clock, cfg, newDB, mode);
+}
+
+bool
+Application::modeHasOverlay(AppMode m)
+{
+    switch (m)
+    {
+    case AppMode::RUN_LIVE_NODE:
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/main/Application.cpp]: Ah, I overestimated the awesome power of the "modern C++ `enum class`" construct: it warns on non-exhaustive now (better than old-style `enum`) but still doesn't hard-error like a reasonable language construct. Fixed.

File: src/main/Application.h [hunk 7]
```diff
@@ -157,6 +181,33 @@ class Application
...
+        REPLAY_HISTORY_FOR_METADATA
+    };
+
+    // AppModes should be accessed using either a narrow centralized query like
+    // one of the following static functions, or an exhaustive switch on the
+    // enum. These options give the least chance of missing or combining cases
+    // incorrectly.
+    static bool modeHasOverlay(AppMode m);
+    static bool modeHasDatabase(AppMode m);
+    static bool modePublishesHistory(AppMode m);
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/main/Application.h]: I expect we'll need a few more such mode judgements but I didn't have a clear enough idea of which ones we'd need. I'm going to leave this as-is for now since this PR isn't switching bucket behaviour (yet).

File: src/main/ApplicationImpl.h [hunk 8]
```diff
@@ -122,6 +123,7 @@ class ApplicationImpl : public Application
   private:
     VirtualClock& mVirtualClock;
     Config mConfig;
+    AppMode mAppMode;
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/main/ApplicationImpl.h]: Good catch, fixed.

File: src/main/Application.h [hunk 9]
```diff
@@ -175,6 +226,11 @@ class Application
     virtual std::string getStateHuman() const = 0;
     virtual bool isStopping() const = 0;
 
+    // Gets the current Application mode. This is a constant established
+    // when the Application is constructed and will not change while it
+    // runs.
+    virtual AppMode getMode() const = 0;
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/main/Application.h]: Sure. Fixed.

File: src/main/Application.cpp [hunk 10]
```diff
@@ -37,8 +37,51 @@ validateNetworkPassphrase(Application::pointer app)
...
+        return true;
+    case AppMode::RELAY_LIVE_TRAFFIC:
+        return false;
+    case AppMode::REPLAY_HISTORY_FOR_METADATA:
+        return false;
+    }
+}
+
+bool
+Application::modePublishesHistory(AppMode m)
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/main/Application.cpp]: Fixed.

File: src/history/HistoryManagerImpl.cpp [hunk 11]
```diff
@@ -222,6 +229,11 @@ HistoryManagerImpl::getMaxLedgerQueuedToPublish()
 bool
 HistoryManagerImpl::maybeQueueHistoryCheckpoint()
 {
+    if (!Application::modePublishesHistory(mApp.getMode()))
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/history/HistoryManagerImpl.cpp]: Good catch, fixed.

File: src/main/Application.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (graydon) [reply src/main/Application.h]: hm, we have such interfaces as static helpers (Application::modeHasFoo(AppMode)); I figured it might be helpful to have the flexibility to switch or use the helpers in different contexts (and/or reify the mode as a value you might want to store / pass around). I guess these are hypotheticals though; can change if you feel strongly!

File: src/main/ApplicationImpl.h [hunk 12]
```diff
@@ -156,6 +158,14 @@ class ApplicationImpl : public Application
     std::unique_ptr<StatusManager> mStatusManager;
     std::unique_ptr<LedgerTxnRoot> mLedgerTxnRoot;
 
+    // These two exist for use in AppMode::REPLAY_HISTORY_FOR_METADATA only: the
+    // InMemoryLedgerTxnRoot is a stub AbstractLedgerTxnParent that refuses all
+    // commits and answers null to all queries; then a "never-committing"
+    // sub-LedgerTxn is constructed beneath it that serves as the "effective"
+    // in-memory root transaction.
+    std::unique_ptr<InMemoryLedgerTxnRoot> mInMemoryLedgerTxnRoot;
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/main/ApplicationImpl.h]: Oh! Wow, I literally did not know unique_ptr would accept an abstract class as its type parameter. Learn something every day. Done!

File: src/ledger/LedgerTxn.h [hunk 13]
```diff
@@ -391,6 +391,50 @@ class AbstractLedgerTxnParent
     // or if the corresponding LedgerEntry has been erased.
     virtual std::shared_ptr<LedgerEntry const>
     getNewestVersion(LedgerKey const& key) const = 0;
+
+    virtual uint64_t
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/ledger/LedgerTxn.h]: I'm reasonably comfortable changing them to "implemented as throw on anything other than a root LedgerTxn". Is that ok? They're never used outside of that context, and they seem pointlessly tricky to try to implement "for completeness" on other LedgerTxn subclasses.

File: src/ledger/LedgerManagerImpl.h (hunk 1 above)
Comment (graydon) [reply src/ledger/LedgerManagerImpl.h]: Ah, good point. Fixed.

File: src/ledger/LedgerManagerImpl.cpp [hunk 14]
```diff
@@ -970,6 +997,45 @@ LedgerManagerImpl::deleteOldEntries(Database& db, uint32_t ledgerSeq,
     txscope.commit();
 }
 
+void
+LedgerManagerImpl::setupLedgerCloseMetaStream()
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/ledger/LedgerManagerImpl.cpp]: Oh wow! I don't know how calling that got dropped from the diff, added back!

I guess I need to, uh, make some tests :)

File: src/ledger/LedgerManagerImpl.cpp [hunk 15]
```diff
@@ -888,12 +896,16 @@ LedgerManagerImpl::closeLedger(LedgerCloseData const& ledgerData)
             LedgerTxn ltxUpgrade(ltx);
             Upgrades::applyTo(lupgrade, ltxUpgrade);
 
-            auto ledgerSeq = ltxUpgrade.loadHeader().current().ledgerSeq;
+            LedgerEntryChanges changes = ltxUpgrade.getChanges();
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/ledger/LedgerManagerImpl.cpp]: Hm, I think somehow I thought it was necessary. Trying without.

File: src/ledger/LedgerManagerImpl.cpp [hunk 16]
```diff
@@ -888,12 +896,16 @@ LedgerManagerImpl::closeLedger(LedgerCloseData const& ledgerData)
...
             // Note: Index from 1 rather than 0 to match the behavior of
             // storeTransaction and storeTransactionFee.
-            Upgrades::storeUpgradeHistory(getDatabase(), ledgerSeq, lupgrade,
-                                          ltxUpgrade.getChanges(),
-                                          static_cast<int>(i + 1));
+            if (Application::modeHasDatabase(mApp.getMode()))
+            {
+                auto ledgerSeq = ltxUpgrade.loadHeader().current().ledgerSeq;
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/ledger/LedgerManagerImpl.cpp]: I think I just moved a few too many things around in the process of getting it to work in both with- and without-db modes. Will try to reorganize closer to how it was.

[the remark this answers is not in the document]
Comment (graydon) [reply src/ledger/LedgerManagerImpl.cpp]: Motivation shows up in later patch in the PR (00e8866b73abc478a6b3d78cdd09edbde161b551) where we need to have the changes pulled out early (and separately) in order to optionally send them to either or both of the stream and the database.

[the remark this answers is not in the document]
Comment (graydon) [reply src/ledger/LedgerManagerImpl.cpp]: (That said, you're right it doesn't quite work as written, will fix)

File: src/ledger/LedgerManagerImpl.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (graydon) [reply src/ledger/LedgerManagerImpl.cpp]: see comment below.

File: src/herder/TxSetFrame.cpp (hunk 2 above)
Comment (graydon) [reply src/herder/TxSetFrame.cpp]: Mm, ok. I'll call `sortForHash` explicitly. Sigh. This class is .. kinda a mess.

**Comment by graydon** [comment]
Rebased, addressed all review comments, added test.

**Comment by graydon** [comment]
Repushed with new test that runs replay mode and fixes a couple of the throwing methods on ledgertxn to call through to their parent so they can be used in that mode.

File: src/history/HistoryManagerImpl.cpp [hunk 17]
```diff
@@ -147,6 +148,12 @@ HistoryManagerImpl::logAndUpdatePublishStatus()
 size_t
 HistoryManagerImpl::publishQueueLength() const
```
Comment (marta-lokhova) [inline src/history/HistoryManagerImpl.cpp]: General question on this class: it looks like right now the choice between throwing and simulating function behavior is sort of arbitrary. For example, we allow to "publish" queued history, or to check publish queue length, but throw if we're trying to queue something up. It'd be nice to have consistent behavior across functions (note for the future: maybe at some point we can even not spin up HistoryManager at all when in "publishing disabled" mode, i think it's centrally controlled in LedgerManager anyway)

File: src/main/Application.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/main/Application.h]: Yeah I agree, the code is littered with `Application::modeHasDatabase(mApp.getMode())` . 
Also, something like `mApp.hasDatabase()` is easier to read

File: src/ledger/LedgerCloseMetaStream.cpp [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+
+#include "ledger/LedgerCloseMetaStream.h"
+#include "util/GlobalChecks.h"
+#include "util/Logging.h"
+#include "xdrpp/marshal.h"
+
+namespace stellar
+{
+
+LedgerCloseMetaStream::LedgerCloseMetaStream(asio::io_context& ctx,
```
Comment (MonsieurNicolas) [inline src/ledger/LedgerCloseMetaStream.cpp]: Do we gain anything from the extra level of buffering to write to the pipe?

For the "rebuild the meta" scenario, the "dumb" implementation would just synchronously write on the main thread (blocking). This would not require any throttling logic, would not do an extra copy of the data into buffers, and the only assumption would be that the reader would consume data as it gets pushed onto the pipe (and as we copy less, we may have a net perf increase).

If it's to make the "online" version work, it's probably premature as this design doesn't handle failures anyways (and this PR doesn't even deal with shutdown properly if not run with `catchup`).

File: src/ledger/test/LedgerCloseMetaStreamTests.cpp [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+#include "lib/catch.hpp"
+#include "main/Application.h"
+#include "main/ApplicationUtils.h"
+#include "test/TestUtils.h"
+#include "test/test.h"
+#include "xdr/Stellar-ledger.h"
+
+using namespace stellar;
+
+#ifndef _WIN32
```
Comment (MonsieurNicolas) [inline src/ledger/test/LedgerCloseMetaStreamTests.cpp]: Is there a reason that those tests are not written cross platform?

File: src/ledger/test/LedgerCloseMetaStreamTests.cpp [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+    cfg.METADATA_OUTPUT_FILE_DESCRIPTOR = fd;
+    VirtualClock clock;
+    auto app = createTestApplication(clock, cfg);
+    app->start();
+    while (app->getLedgerManager().getLastClosedLedgerNum() < 10)
+    {
+        clock.crank(true);
+    }
+    Hash hash = app->getLedgerManager().getLastClosedLedgerHeader().hash;
+    while (!app->getLedgerManager().metadataBufferEmpty())
```
Comment (MonsieurNicolas) [inline src/ledger/test/LedgerCloseMetaStreamTests.cpp]: you should not need to wait for the meta buffer to be empty (that seems to be missing logic in the app shutdown code: the "wrapper process" would not be able to do this) - regardless, I don't think we should support `LIVE_NODE` in this PR as there are too many unknowns around it

File: src/main/CommandLine.cpp [hunk 21]
```diff
@@ -1024,6 +1050,9 @@ handleCommandLine(int argc, char* const* argv)
           runSignTransaction},
          {"upgrade-db", "upgade database schema to current version",
           runUpgradeDB},
+         {"replay-history-for-metadata",
```
Comment (MonsieurNicolas) [inline src/main/CommandLine.cpp]: this should be an option for the `catchup` sub-command instead of being exposed at the top level like this

File: src/main/Config.cpp [hunk 22]
```diff
@@ -1226,6 +1241,14 @@ Config::validateConfig(ValidationThresholdLevels thresholdLevel)
                                   UNSAFE_QUORUM ? 1 : 51);
         throw std::invalid_argument("Invalid QUORUM_SET");
     }
+
+    if (METADATA_OUTPUT_FILE_DESCRIPTOR != -1 &&
+        METADATA_OUTPUT_NAMED_PIPE != "")
+    {
+        LOG(FATAL) << "Can only set one of METADATA_OUTPUT_NAMED_PIPE and"
```
Comment (MonsieurNicolas) [inline src/main/Config.cpp]: we could also only have one setting that can be either a path or an integer?

Comment (graydon) [reply src/main/Config.cpp]: I think this is not ideal: it's weird to the user (precludes integer-valued filenames, granted, unlikely) and requires weird implementation (a config-value visitor I guess? something that precludes accessing the string as a string when it's "really" an integer..)

File: src/ledger/test/LedgerCloseMetaStreamTests.cpp (hunk 19 above)
Comment (graydon) [reply src/ledger/test/LedgerCloseMetaStreamTests.cpp]: I initially thought I'd be using anonymous `pipe()` rather than a file-backed fd. I suppose the file-backed fd should work on windows too. I'll give it a try.

File: src/history/HistoryManagerImpl.cpp [hunk 23]
```diff
@@ -147,6 +148,12 @@ HistoryManagerImpl::logAndUpdatePublishStatus()
...
 {
+    if (!Application::modePublishesHistory(mApp.getMode()) ||
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/history/HistoryManagerImpl.cpp]: As it happens we won't be able to publish history if we don't have a database -- the checkpoints all bounce off a database table that's not replicated in the no-database case, we just don't even store it.

So I could collapse this into `modeHasDatabase`, but we need to not-even-try to publish history when we're in a no-database mode.

[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/history/HistoryManagerImpl.cpp]: so maybe bomb out on startup if there is a writeable archive but we're configured with no database?

File: src/herder/TxSetFrame.cpp [hunk 24]
```diff
@@ -512,6 +513,7 @@ TxSetFrame::getTotalFees(LedgerHeader const& lh) const
 void
 TxSetFrame::toXDR(TransactionSet& txSet)
 {
+    sortForHash();
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/herder/TxSetFrame.cpp]: It is because `TxSetFrame` and friends are a mess. `AbstractTxSetFrame` (which is all we have a pointer to in `LedgerManagerImpl`) does not have a `sortForHash` method, and shouldn't because `SimulationTxSetFrame` has only `const` members (this is a good design I want to keep).

In contrast, `SimulationTxSetFrame::toXDR` actually constructs, sorts-for-hash, and calls `toXDR` on a temporary `TxSet`, which it then throws out. I could theoretically do something like this in `TxSetFrame::toXDR` as well -- at least the construction-and-sorting part, the delegation has to bottom out with an xdrpp call somewhere! -- and remove or change the other call sites. Not sure what you prefer.

I should also point out that the not-very-well-named `TxSetFrame::sortForApply` doesn't actually sort-in-place either: it produces a new sorted vector for application. I'd be happy to try moving the `sortForHash` interface to this sort of shape too, but I wasn't sure how much deeper the pile of yak hair should get in this PR. Happy to go deeper if you like.

Basically `TxSetFrame` is a mutable/stateful type that's half way through migrating to an immutable/stateless type. I was trying to fit into that existing situation without knocking too many other things over.

File: src/herder/TxSetFrame.cpp [hunk 25]
```diff
@@ -512,6 +513,7 @@ TxSetFrame::getTotalFees(LedgerHeader const& lh) const
 void
 TxSetFrame::toXDR(TransactionSet& txSet)
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/herder/TxSetFrame.cpp]: Not .. entirely. Depends how much mess I make / clean up in the process. See comment below re: nicolas' question about the same thing.

File: src/ledger/LedgerCloseMetaStream.cpp (hunk 18 above)
Comment (graydon) [reply src/ledger/LedgerCloseMetaStream.cpp]: Done.

File: src/ledger/test/LedgerCloseMetaStreamTests.cpp (hunk 20 above)
Comment (graydon) [reply src/ledger/test/LedgerCloseMetaStreamTests.cpp]: Fixed in process of making it synchronous.

File: src/history/HistoryManagerImpl.cpp (hunk 17 above)
Comment (graydon) [reply src/history/HistoryManagerImpl.cpp]: Unfortunately HistoryManager provides also the interface we use to do ledger-range calculations (based on whether or not we're in accelerated-time mode) and a few other query-history helpers, so there are several dozen call sites even in not-publishing mode. Again, I'm not strictly opposed to eliminating these / trying to move to a no-history-manager mode, but it involves yet more code motion and the PR is already pretty big.

File: src/main/ExternalQueue.cpp [hunk 26]
```diff
@@ -25,6 +25,7 @@ string ExternalQueue::kSQLCreateStatement =
 
 ExternalQueue::ExternalQueue(Application& app) : mApp(app)
 {
+    releaseAssertOrThrow(Application::modeHasDatabase(mApp.getMode()));
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/main/ExternalQueue.cpp]: Fixed.

File: src/main/ApplicationImpl.cpp [hunk 27]
```diff
@@ -333,6 +360,15 @@ ApplicationImpl::~ApplicationImpl()
     {
         mProcessManager->shutdown();
     }
+    if (!modeHasDatabase(mAppMode))
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/main/ApplicationImpl.cpp]: Because at least in _this_ PR we're treating "no-database" mode as meaning "form buckets while you're running to check hashes, but don't keep buckets around when you shut down". This is debatable.

File: src/main/CommandLine.cpp (hunk 21 above)
Comment (graydon) [reply src/main/CommandLine.cpp]: Heh. The comment I put in for modes specifically says it's for when cases where it does something meaningfully different than an existing subcommand. I think in this case it really does! `catchup` modifies local state; `replay-history-for-metadata` is stateless, idempotent. From the user's perspective it's quite different than catchup; it just happens to do catchup internally. No?

File: src/history/HistoryManagerImpl.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (graydon) [reply src/history/HistoryManagerImpl.cpp]: switched to !modeHasDatabase() in all cases, removed modePublishesHistory()

[the remark this answers is not in the document]
Comment (graydon) [reply src/history/HistoryManagerImpl.cpp]: done.

File: src/main/Application.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (graydon) [reply src/main/Application.h]: fixed in all cases (though delegating to the static methods, which are still there for querying a mode _before_ the application is built).

File: src/main/Config.cpp (hunk 22 above)
Comment (graydon) [reply src/main/Config.cpp]: Fixed: went with "pathname which can be a pipe or a file, or 'fd:N' which opens the N file descriptor". User can deal with living without files named "fd:N".

File: src/main/CommandLine.cpp (hunk 21 above)
Comment (graydon) [reply src/main/CommandLine.cpp]: fixed, made into a `catchup` argument `--replay-in-memory`.

File: src/ledger/test/LedgerCloseMetaStreamTests.cpp (hunk 19 above)
Comment (graydon) [reply src/ledger/test/LedgerCloseMetaStreamTests.cpp]: windows tests work now too. and confirmed named pipes still work.

**Comment by graydon** [comment]
Ok I believe this now addresses all review comments so far. Need to squash & rebase but it's (thankfully) getting quite small!

**Comment by graydon** [comment]
Squashed and formatted. Running tests now. Ought to be nearly ready.

File: src/history/HistoryManagerImpl.cpp [hunk 28]
```diff
@@ -317,6 +333,11 @@ HistoryManagerImpl::publishQueuedHistory()
     }
 #endif
 
+    if (!mApp.modeHasDatabase())
```
Comment (MonsieurNicolas) [inline src/history/HistoryManagerImpl.cpp]: meta question here...

I think we could simplify (and remove more of those `if (!mApp.modeHasDatabase())`) calls:
what if we were instead setting up an in memory sqlite database, that way code that deals with tables that are low churn like `PersistentState` would not have to be changed.
We could also not create the expensive tables at all, but we can also just check in tests that we don't populate them (and a bug in that front would be trivial to spot as we would have large memory usage).

File: src/main/ApplicationImpl.cpp [hunk 29]
```diff
@@ -333,6 +363,18 @@ ApplicationImpl::~ApplicationImpl()
...
+    {
+        // Note: BucketManager::dropAll will delete the $BUCKETDIR/tmp which is
+        // where process and history manager write _their_ temp files to, so we
+        // make sure to tear them down first here.
+        mProcessManager = nullptr;
```
Comment (MonsieurNicolas) [inline src/main/ApplicationImpl.cpp]: should this be done after `joinAllThreads` below?

File: src/herder/TxSetFrame.cpp (hunk 24 above)
[the remark this answers is not in the document]
Comment (graydon) [reply src/herder/TxSetFrame.cpp]: fixed to just assert that it's sorted

File: docs/stellar-core_example.cfg [hunk 30]
```diff
@@ -384,6 +384,27 @@ DISABLE_XDR_FSYNC=false
...
+# at startup and synchronously streamed-to during both catchup and
+# live ledger-closing.
+#
+# Streams may be specified either as an existing pathname (typically a
+# named FIFO on POSIX or a named pipe on Windows) or a string of the
+# form "fd:N" for some integer N which, on POSIX, specifies the
+# existing open file descriptor N inherited by the process (for
+# example to write to an anonymous pipe). Anonymous file descriptor
+# streams are not supported on Windows (anonymous pipes do not work
+# with the ASIO library used in stellar-core).
```
[the remark this answers is not in the document]
Comment (graydon) [reply docs/stellar-core_example.cfg]: fixed

File: src/history/HistoryManagerImpl.cpp [hunk 31]
```diff
@@ -222,6 +230,11 @@ HistoryManagerImpl::getMaxLedgerQueuedToPublish()
 bool
 HistoryManagerImpl::maybeQueueHistoryCheckpoint()
 {
+    if (!mApp.modeHasDatabase())
```
[the remark this answers is not in the document]
Comment (graydon) [reply src/history/HistoryManagerImpl.cpp]: fixed

File: src/history/HistoryManagerImpl.cpp (hunk 28 above)
Comment (graydon) [reply src/history/HistoryManagerImpl.cpp]: Possible! But will defer this question to a later PR as discussed in chat.

File: src/main/ApplicationImpl.cpp (hunk 29 above)
Comment (graydon) [reply src/main/ApplicationImpl.cpp]: oh, good catch. sure, fixed.

**Comment by MonsieurNicolas** [comment]
r+ 6eff08a8f6e7053be8c1dbcf95e375c551a2f3cb
